### PR TITLE
New protocol layer for blob, datalake and queue generated by CodeGen

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-blobs/swagger/README.md [hunk 1]
```diff
@@ -0,0 +1,1361 @@
...
+      $["200"].schema = {
+        "x-ms-client-name": "GetBlockListResult",
+        "xml": {"name": "BlockList"},
+        "type": "object",
+        "properties": {
+          "ETag": {"type": "string", "format": "etag", "x-ms-xml": {"name": ""}, "description": "The ETag contains a value that you can use to perform operations conditionally. If the request version is 2011-08-18 or newer, the ETag value will be in quotes."},
+          "LastModified": {"type": "string", "format": "date-time-rfc1123", "x-ms-xml": {"name": ""}, "description": "Returns the date and time the container was last modified. Any operation that modifies the blob, including an update of the blob's metadata or properties, changes the last-modified time of the blob."},
+          "BlobSize": {"type": "integer", "format": "int64", "x-ms-client-default": "0", "x-ms-xml": {"name": ""}, "description": "Size of the blob in bytes."},
+          "CommittedBlocks": {"type": "array", "items": {"$ref": "#/definitions/Block"}, "description": "List of committed blocks."},
+          "UncommittedBlocks": {"type": "array", "items": {"$ref": "#/definitions/Block"}, "description": "List of uncommitted blocks."}
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/swagger/README.md:1358]: what is this for?

**Review by vhvb1989 — COMMENTED** [review]
Tests are failing on CI.  That means either:

- Current test recordings need to be updated since the network calls from the protocol layer have changed.  To update recordings, set env var `AZURE_TEST_MODE=RECORD` and run each failing test ( no not run all storage tests to avoid new recordings for all tests! ).  Set env var `AZURE_TEST_MOE=PLAYBACK` and re-run the tests to confirm the new recording is good for the test.
- If same recording should work, then probably the test case is no longer valid for the new implementation and test need to be updated as well as the recording.
-

**Comment by vhvb1989** [comment]
Here's a PR to your branch with recordings updated:
https://github.com/Jinming-Hu/azure-sdk-for-cpp/pull/1

Looks like some networks calls changed from `GET` to `HEAD`.  So the recordings needed to be updated.

File: sdk/storage/azure-storage-blobs/swagger/README.md (hunk 1 above)
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/swagger/README.md:1358]: I am assuming your question is about the whole file, in general.

This is a readme transform file that modifies the swagger before being passed-in to the modeler 4 component of autorest (which is then passed into the language plugin that generates C++). This way, the C++ language plugin/extension can remain agnostic of "tweaks" or corrections we need to make to service specific swagger.

Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/swagger/README.md:1358]: If your question is specific to these particular lines your comment highlighted that are making the modifications, @Jinming-Hu can talk more to that.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/swagger/README.md:1358]: These lines are used to change the return type of function `BlockBlobClient::GetBlockList`. In the original swagger file, as you can see [here](https://github.com/Jinming-Hu/azure-storage-api-specs/blob/main/Microsoft.BlobStorage/preview/2020-02-10/blob.json#L7076) and [here](https://github.com/Jinming-Hu/azure-storage-api-specs/blob/73a91e111c677d392d8459f6e06ac1b7f2ff4a3e/Microsoft.BlobStorage/preview/2020-02-10/blob.json#L11240), the return type is

```C++
struct XXX {
  std::vector<BlockList>CommittedBlocks;
  std::vector<BlockList> UncommittedBlocks;
};
```

but in the existing code, the return type is

https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/protocol/blob_rest_client.hpp#L1918

**Comment by vhvb1989** [comment]
@ahsonkhan is this using your c++ auto-rest?  It looks like the code is not clang-format compliance and needs update.
Otherwise, if it's not from auto-gen, this needs manual formatting update

**Comment by Jinming-Hu** [comment]
> @ahsonkhan is this using your c++ auto-rest? It looks like the code is not clang-format compliance and needs update. Otherwise, if it's not from auto-gen, this needs manual formatting update

That is convenience layer code. I've fixed it.

**Review by vhvb1989 — APPROVED** [review]

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 2]
```diff
@@ -1000,7 +1000,22 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+       * is less than or equal to this value.
+       */
+      Azure::Nullable<int64_t> IfMaxSizeLessThanOrEqual;
+
+      /**
+       * @brief Ensures that the AppendBlock operation succeeds only if the append position is equal
+       * to this value.
+       */
+      Azure::Nullable<int64_t> IfAppendPositionEqual;
+    } AccessConditions;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:1018]: Breaking change

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs.hpp [hunk 3]
```diff
@@ -17,3 +17,4 @@
 #include "azure/storage/blobs/block_blob_client.hpp"
 #include "azure/storage/blobs/dll_import_export.hpp"
 #include "azure/storage/blobs/page_blob_client.hpp"
+#include "azure/storage/common/storage_exception.hpp"
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs.hpp:20]: Why do we need this here, especially now?

We should avoid exposing headers from other packages from within the simplified main.hpp header for the package.

For example: blobs.hpp shouldn't contain headers from storage common, and only contain all the main headers we want customers to include from blobs. If those blobs header transitively include other headers, that's fine, but they shouldn't be in the main simplified header.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues.hpp [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 #pragma once
 
+#include "azure/storage/common/storage_exception.hpp"
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-queues/inc/azure/storage/queues.hpp:11]: Same issue here, and possibly elsewhere. Keep queues.hpp simplified header self-contained for headers in that package only.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs.hpp (hunk 3 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs.hpp:20]: In the old protocol layer code, function definitions were in .hpp file, so `StorageException` was indirectly exposed to customers. Customers could use this it without including the header.

But in the new protocol layer code, function definitions were moved to .cpp file, thus customers don't have access to the exception unless they explicitly include `storage_exception.hpp`. Unfortunately, storage exception is something that's being used by a lot of customers, and it's not easy for them to figure out that they should include another header after upgrading to a new version for the existing code to keep working. This is like a breaking change. So I added `storage_exception.hpp` to `blobs.hpp`, `queues.hpp` and `datalake.hpp` as a mitigation.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues.hpp (hunk 4 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues.hpp:11]: duplicate, will discuss in another thread

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs.hpp (hunk 3 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs.hpp:20]: I know this is not ideal. But I cannot come up with a better solution.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp (hunk 2 above)
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp]: Do we still have this breaking change?

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp]: No, no breaking changes will be made in this release.

**Comment by Jinming-Hu** [comment]
@ahsonkhan 

We can use APIView to compare the difference.

**Review by katmsft — APPROVED** [review]

**Pull request merged**